File: Assets/Bundles/Client/Resources/TS/Controllers/Global/Camera/CameraReferences.lua
-- Compiled with unity-ts v2.1.0-75
local WorldAPI = require("Shared/TS/VoxelWorld/WorldAPI").WorldAPI
local Mouse = require("Shared/TS/UserInput/init").Mouse
local CameraReferences
do
	CameraReferences = setmetatable({}, {
		__tostring = function()
			return "CameraReferences"
		end,
	})
	CameraReferences.__index = CameraReferences
	function CameraReferences.new(...)
		local self = setmetatable({}, CameraReferences)
		return self:constructor(...) or self
	end
	function CameraReferences:constructor()
		self.CameraBundleKey = "CameraRig"
		self.mouse = Mouse.new()
		if CameraReferences._instance then
			error("TRYING TO INITIALIZE SINGLETON THAT ALREADY EXISTS: CameraReferences")
			return nil
		end
		CameraReferences._instance = self
		-- Get Camera references
		local references = GameObjectReferences:GetReferences(self.CameraBundleKey)
		self.mainCamera = references:GetValue("Cameras", "MainCamera")
		self.fpsCamera = references:GetValue("Cameras", "FPSCamera")
	end
	function CameraReferences:Instance()
		if not CameraReferences._instance then
			CameraReferences.new()
		end
		return CameraReferences._instance
	end
	function CameraReferences:RaycastVoxelFromCamera(distance)
		local ray = self:GetRayFromCamera(distance)
		return WorldAPI:GetMainWorld():RaycastVoxel(ray.origin, ray.direction, distance)
	end
	function CameraReferences:RaycastPhysicsFromCamera(distance, layerMask)
		local ray = self:GetRayFromCamera(distance)
		return Physics:EasyRaycast(ray.origin, ray.direction, 40, layerMask)
	end
	function CameraReferences:GetRayFromCamera(distance)
		local ray = Camera.main:ScreenPointToRay(self.mouse:GetLocation())
		local _direction = ray.direction
		local _distance = distance
		ray.direction = _direction * _distance
		return ray
	end
end
return {
	CameraReferences = CameraReferences,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Client/Controllers/Global/Camera/CameraReferences.ts
import { WorldAPI } from "../../../../Shared/VoxelWorld/WorldAPI";
import { Mouse } from "../../../../Shared/UserInput";

export class CameraReferences {
	private readonly CameraBundleKey = "CameraRig";

	private static _instance: CameraReferences;
	public static Instance(): CameraReferences {
		if (!CameraReferences._instance) {
			new CameraReferences();
		}
		return CameraReferences._instance;
	}

	private mouse = new Mouse();
	public readonly mainCamera: Camera;
	public readonly fpsCamera: Camera;

	public constructor() {
		if (CameraReferences._instance) {
			error("TRYING TO INITIALIZE SINGLETON THAT ALREADY EXISTS: CameraReferences");
			return;
		}
		CameraReferences._instance = this;

		//Get Camera references
		let references = GameObjectReferences.GetReferences(this.CameraBundleKey);
		this.mainCamera = references.GetValue<Camera>("Cameras", "MainCamera");
		this.fpsCamera = references.GetValue<Camera>("Cameras", "FPSCamera");
	}

	public RaycastVoxelFromCamera(distance: number) {
		const ray = this.GetRayFromCamera(distance);
		return WorldAPI.GetMainWorld().RaycastVoxel(ray.origin, ray.direction, distance);
	}

	public RaycastPhysicsFromCamera(distance: number, layerMask?: number) {
		const ray = this.GetRayFromCamera(distance);
		return Physics.EasyRaycast(ray.origin, ray.direction, 40, layerMask);
	}

	public GetRayFromCamera(distance: number) {
		const ray = Camera.main.ScreenPointToRay(this.mouse.GetLocation());
		ray.direction = ray.direction.mul(distance);
		return ray;
	}
}
